File: src/ClaudeStatusBar/Model/StateMachine.cs
namespace ClaudeStatusBar.Model;

/// <summary>
/// Raw state table, grace cap and hysteresis, per docs/forecast-and-states.md,
/// "States (per window)". Three independent, directly testable pieces:
///
///   RawStateClassifier - the priority table (Spent > Measuring > DryEarly > Tight > Safe).
///   GraceCap           - the near-reset clamp, applied after hysteresis, every tick.
///   HysteresisHold      - per-window debounce between verdict states, stepped once per
///                         ACCEPTED (novel) observation, never per poll tick or duplicate.
///
/// Clarification applied here (see doc's "Refusal -> Measuring" note): leaving
/// Measuring commits immediately, because Measuring is not a verdict and there
/// is nothing to debounce -- hysteresis only ever arbitrates between the three
/// verdict states (Safe/Tight/DryEarly). Spent, grace and rollover all bypass
/// hysteresis, as the doc says.
/// </summary>
public static class RawStateClassifier
{
    public static QuotaState Classify(double usedPct, ForecastResult forecast, double windowMinutes, bool refused)
    {
        // Decision 15: Spent only at P >= 100 (not 99.5) -- confirmed exhaustion, not "almost".
        // 97-99.9% still reaches DryEarly on its own via the existing RemainingPct <= 3 clause below.
        if (usedPct >= 100.0) return QuotaState.Spent;
        if (refused) return QuotaState.Measuring;

        double sLo = Math.Max(0.05 * windowMinutes, 10.0);
        double sHi = Math.Max(0.08 * windowMinutes, 15.0);

        if (forecast.ShortfallMinutes >= sLo || forecast.RemainingPct <= 3.0) return QuotaState.DryEarly;
        if (forecast.ShortfallMinutes > 0.0 || forecast.SlackMinutes <= sHi || forecast.RemainingPct <= 10.0) return QuotaState.Tight;
        return QuotaState.Safe;
    }
}

public static class GraceCap
{
    /// <summary>
    /// Decision 2: grace must never turn a predicted cutoff into Safe. The only
    /// remaining cap is t_reset &lt;= 15 min, which caps DryEarly at Tight (no red
    /// right before a reset that will clear it anyway); the old t_reset &lt;= 5 min
    /// -&gt; Safe rule is removed entirely. Spent is never capped. A cap can only
    /// ever lower severity, never raise it.
    /// </summary>
    public static QuotaState Apply(QuotaState state, double minutesToReset)
    {
        if (state == QuotaState.Spent) return state;
        if (minutesToReset <= 15.0) return (QuotaState)Math.Min((int)state, (int)QuotaState.Tight);
        return state;
    }
}

/// <summary>
/// Per-window hysteresis. Step() must be called only for ACCEPTED (novel)
/// observations -- decision 11: duplicate replies and poll ticks are not
/// evidence, and a failed poll neither counts as evidence nor resets pending
/// evidence (the caller simply never calls Step() for either). Committed
/// reflects the result on every subsequent read until the next Step()/
/// CommitFromMeasuring().
///
/// Round 2 (finding 2, decision 2): all persistence -- LastChangeMono, the
/// cooldown, and the de-escalation minimum-elapsed-time gate -- is monotonic
/// (long milliseconds from Environment.TickCount64), not UTC. A poll response
/// this class ever sees for Step() is always live (QuotaModel never steps
/// hysteresis from WarmStart's replay path), so its monoMs is always a
/// trustworthy live reading -- no UTC/mono reconciliation is needed here the
/// way WindowTracker needs one for out-of-order detection.
///
/// Escalation (decision 11): commits to the LOWEST severity that is still
/// above Committed, seen across the last _escalateEvidenceCount accepted
/// observations -- not to exact repeated agreement on one target state. This
/// is what lets an alternating Tight/DryEarly escalate at all (both qualify
/// as "above Committed"; the old exact-match rule could stall on Safe
/// forever, per review finding 11).
///
/// De-escalation needs _deescalateEvidenceCount consecutive accepted
/// observations agreeing on the same lower state, AND at least
/// _deescalateMinElapsedMs real time since the first of them -- replacing the
/// old poll-count-only thresholds (6/20), which scaled with polling policy
/// instead of wall-clock confidence.
///
/// The original "at most one committed change per cooldown" rule still gates
/// both directions.
/// </summary>
public sealed class HysteresisHold
{
    readonly int _escalateEvidenceCount;
    readonly int _deescalateEvidenceCount;
    readonly long _deescalateMinElapsedMs;
    readonly long _cooldownMs;

    readonly List<QuotaState> _escalateHistory = new();
    QuotaState? _pendingDeescalateState;
    int _pendingDeescalateCount;
    long? _pendingDeescalateSinceMono;

    public QuotaState Committed { get; private set; } = QuotaState.Measuring;
    public long? LastChangeMono { get; private set; }

    public HysteresisHold(bool weekly)
    {
        _escalateEvidenceCount = weekly ? 4 : 2;
        _deescalateEvidenceCount = weekly ? 4 : 3;
        _deescalateMinElapsedMs = (long)(weekly ? TimeSpan.FromHours(2) : TimeSpan.FromMinutes(10)).TotalMilliseconds;
        _cooldownMs = (long)(weekly ? TimeSpan.FromMinutes(5) : TimeSpan.FromSeconds(90)).TotalMilliseconds;
    }

    /// <summary>Call only for an accepted (novel) observation -- see the type doc comment.</summary>
    public QuotaState Step(QuotaState raw, long monoMs)
    {
        // Spent, Measuring (entering or leaving) and rollover (which always forces raw ==
        // Measuring while its gate holds) all bypass hysteresis: commit immediately.
        if (raw == QuotaState.Spent || raw == QuotaState.Measuring
            || Committed == QuotaState.Measuring || Committed == QuotaState.Spent)
        {
            CommitImmediately(raw, monoMs);
            return Committed;
        }

        if (raw == Committed)
        {
            ResetPendingEvidence();
            return Committed;
        }

        bool cooldownElapsed = LastChangeMono is not { } lastChange || (monoMs - lastChange) >= _cooldownMs;
        bool escalating = (int)raw > (int)Committed;

        if (escalating)
        {
            _pendingDeescalateState = null;
            _pendingDeescalateCount = 0;
            _pendingDeescalateSinceMono = null;

            _escalateHistory.Add(raw);
            if (_escalateHistory.Count > _escalateEvidenceCount) _escalateHistory.RemoveAt(0);

            bool allAboveCommitted = _escalateHistory.Count == _escalateEvidenceCount
                && _escalateHistory.All(s => (int)s > (int)Committed);
            if (allAboveCommitted && cooldownElapsed)
            {
                QuotaState target = (QuotaState)_escalateHistory.Min(s => (int)s);
                CommitImmediately(target, monoMs);
            }
        }
        else
        {
            _escalateHistory.Clear();

            if (_pendingDeescalateState != raw)
            {
                _pendingDeescalateState = raw;
                _pendingDeescalateCount = 1;
                _pendingDeescalateSinceMono = monoMs;
            }
            else
            {
                _pendingDeescalateCount++;
            }

            bool enoughEvidence = _pendingDeescalateCount >= _deescalateEvidenceCount;
            bool enoughTime = _pendingDeescalateSinceMono is { } since && (monoMs - since) >= _deescalateMinElapsedMs;
            if (enoughEvidence && enoughTime && cooldownElapsed)
            {
                CommitImmediately(raw, monoMs);
            }
        }

        return Committed;
    }

    /// <summary>
    /// Decision 4 (round 2): leaving Measuring on live time/eligibility alone, with no new
    /// accepted sample -- the rollover/weekly-24h gates lifting purely with elapsed time, or a
    /// WarmStart-seeded window whose only live poll so far was deduped. Bypasses all evidence,
    /// exactly like Step()'s own "leaving Measuring commits immediately" branch, but is safe to
    /// call from a pure evaluation tick (unlike Step(), which would otherwise treat a
    /// non-Measuring Committed's raw as escalate/de-escalate evidence on every 1s tick, in
    /// violation of decision 11). A no-op unless Committed is currently Measuring.
    /// </summary>
    public QuotaState CommitFromMeasuring(QuotaState raw, long monoMs)
    {
        if (Committed != QuotaState.Measuring) return Committed;
        CommitImmediately(raw, monoMs);
        return Committed;
    }

    void ResetPendingEvidence()
    {
        _escalateHistory.Clear();
        _pendingDeescalateState = null;
        _pendingDeescalateCount = 0;
        _pendingDeescalateSinceMono = null;
    }

    void CommitImmediately(QuotaState state, long monoMs)
    {
        Committed = state;
        LastChangeMono = monoMs;
        ResetPendingEvidence();
    }
}

File: src/Mac/Sources/ClaudeQuotaCore/StateMachine.swift
import Foundation

/// Raw state table, grace cap and hysteresis, per docs/forecast-and-states.md,
/// "States (per window)". Ported from `src/Windows/Model/StateMachine.cs`.
///
/// Three independent, directly testable pieces:
///   `RawStateClassifier` — the priority table (Spent > Measuring > DryEarly > Tight > Safe)
///   `GraceCap`           — the near-reset clamp, applied after hysteresis, every tick
///   `HysteresisHold`     — per-window debounce, stepped once per ACCEPTED (novel) observation,
///                          never per poll tick or duplicate
///
/// Leaving Measuring commits immediately, because Measuring is not a verdict and there is
/// nothing to debounce — hysteresis only ever arbitrates between Safe/Tight/DryEarly. Spent,
/// grace and rollover all bypass it.
public enum RawStateClassifier {
    public static func classify(
        usedPct: Double, forecast: ForecastResult, windowMinutes: Double, refused: Bool
    ) -> QuotaState {
        // Decision 15: Spent only at P >= 100 (not 99.5) — confirmed exhaustion, not "almost".
        // 97–99.9 % still reaches DryEarly on its own via the remaining <= 3 clause below.
        if usedPct >= 100.0 { return .spent }
        if refused { return .measuring }

        let sLo = max(0.05 * windowMinutes, 10.0)
        let sHi = max(0.08 * windowMinutes, 15.0)

        if forecast.shortfallMinutes >= sLo || forecast.remainingPct <= 3.0 { return .dryEarly }
        if forecast.shortfallMinutes > 0.0 || forecast.slackMinutes <= sHi || forecast.remainingPct <= 10.0 {
            return .tight
        }
        return .safe
    }
}

public enum GraceCap {
    /// A second live clamp, in the same spirit as the near-reset one below: a committed
    /// **DryEarly whose current forecast shows no shortfall at all** is contradicted by the live
    /// numbers, and presenting it unchanged produces copy that states a falsehood — observed in
    /// the real app as *"⚠ Kvoten tar slut sön kl 00:52"* over *"0 s före reset kl 23:20"*, i.e.
    /// a depletion falling AFTER the reset paired with a shortfall of zero.
    ///
    /// Hysteresis is right to hold the verdict (the pace really was that high, and flapping is
    /// worse), so this does not clear it — it lowers it one step to Tight, which is exactly
    /// "the pace is straining but on current numbers it reaches the reset". Like every cap here
    /// it can only ever lower severity, never raise it, and it never produces Safe: a verdict
    /// that was DryEarly must not become "✓ räcker till reset" on one tick's arithmetic.
    ///
    /// Applies only when there is genuinely no blockage: any shortfall at all, or 3 % or less
    /// remaining, leaves DryEarly exactly as committed.
    public static func applyNoShortfall(_ state: QuotaState, forecast: ForecastResult) -> QuotaState {
        guard state == .dryEarly else { return state }
        guard forecast.shortfallMinutes < 1.0, forecast.remainingPct > 3.0 else { return state }
        return .tight
    }

    /// Decision 2: grace must never turn a predicted cutoff into Safe. The only remaining cap is
    /// `t_reset <= 15 min`, which caps DryEarly at Tight (no red right before a reset that will
    /// clear it anyway); the old `t_reset <= 5 min → Safe` rule is removed entirely. Spent is
    /// never capped, and a cap can only ever lower severity, never raise it.
    public static func apply(_ state: QuotaState, minutesToReset: Double) -> QuotaState {
        if state == .spent { return state }
        if minutesToReset <= 15.0 {
            return QuotaState(rawValue: min(state.rawValue, QuotaState.tight.rawValue)) ?? state
        }
        return state
    }
}

/// Per-window hysteresis. `step` must be called only for ACCEPTED (novel) observations —
/// decision 11: duplicate replies and poll ticks are not evidence, and a failed poll neither
/// counts as evidence nor resets pending evidence (the caller simply never steps for either).
///
/// All persistence — `lastChangeMono`, the cooldown, and the de-escalation minimum-elapsed gate
/// — is **monotonic** milliseconds, not UTC (round-2 decision 2).
///
/// Escalation commits to the LOWEST severity still above `committed`, seen across the last
/// `escalateEvidenceCount` accepted observations — not exact repeated agreement on one target.
/// That is what lets an alternating Tight/DryEarly escalate at all; the old exact-match rule
/// could stall on Safe forever (review finding 11).
///
/// De-escalation needs `deescalateEvidenceCount` consecutive accepted observations agreeing on
/// the same lower state, AND at least `deescalateMinElapsedMs` of real time since the first of
/// them — replacing poll-count-only thresholds, which scaled with polling policy instead of
/// wall-clock confidence.
public final class HysteresisHold {
    private let escalateEvidenceCount: Int
    private let deescalateEvidenceCount: Int
    private let deescalateMinElapsedMs: Int64
    private let cooldownMs: Int64

    private var escalateHistory: [QuotaState] = []
    private var pendingDeescalateState: QuotaState?
    private var pendingDeescalateCount = 0
    private var pendingDeescalateSinceMono: Int64?

    public private(set) var committed: QuotaState = .measuring
    public private(set) var lastChangeMono: Int64?

    public init(weekly: Bool) {
        escalateEvidenceCount = weekly ? 4 : 2
        deescalateEvidenceCount = weekly ? 4 : 3
        deescalateMinElapsedMs = weekly ? 2 * 60 * 60 * 1000 : 10 * 60 * 1000
        cooldownMs = weekly ? 5 * 60 * 1000 : 90 * 1000
    }

    /// Call only for an accepted (novel) observation — see the type doc comment.
    @discardableResult
    public func step(_ raw: QuotaState, monoMs: Int64) -> QuotaState {
        // Spent, Measuring (entering or leaving) and rollover (which forces raw == measuring
        // while its gate holds) all bypass hysteresis: commit immediately.
        if raw == .spent || raw == .measuring || committed == .measuring || committed == .spent {
            commitImmediately(raw, monoMs: monoMs)
            return committed
        }

        if raw == committed {
            resetPendingEvidence()
            return committed
        }

        let cooldownElapsed = lastChangeMono.map { monoMs - $0 >= cooldownMs } ?? true
        let escalating = raw.rawValue > committed.rawValue

        if escalating {
            pendingDeescalateState = nil
            pendingDeescalateCount = 0
            pendingDeescalateSinceMono = nil

            escalateHistory.append(raw)
            if escalateHistory.count > escalateEvidenceCount { escalateHistory.removeFirst() }

            let allAboveCommitted = escalateHistory.count == escalateEvidenceCount
                && escalateHistory.allSatisfy { $0.rawValue > committed.rawValue }
            if allAboveCommitted, cooldownElapsed,
               let lowest = escalateHistory.min(by: { $0.rawValue < $1.rawValue }) {
                commitImmediately(lowest, monoMs: monoMs)
            }
        } else {
            escalateHistory.removeAll()

            if pendingDeescalateState != raw {
                pendingDeescalateState = raw
                pendingDeescalateCount = 1
                pendingDeescalateSinceMono = monoMs
            } else {
                pendingDeescalateCount += 1
            }

            let enoughEvidence = pendingDeescalateCount >= deescalateEvidenceCount
            let enoughTime = pendingDeescalateSinceMono.map { monoMs - $0 >= deescalateMinElapsedMs } ?? false
            if enoughEvidence, enoughTime, cooldownElapsed {
                commitImmediately(raw, monoMs: monoMs)
            }
        }

        return committed
    }

    /// Round-2 decision 4: leaving Measuring on live time/eligibility alone, with no new accepted
    /// sample — the rollover/weekly-24h gates lifting purely with elapsed time, or a
    /// WarmStart-seeded window whose only live poll so far was deduped. Bypasses all evidence,
    /// exactly like `step`'s own "leaving Measuring commits immediately" branch, but is safe to
    /// call from a pure evaluation tick (unlike `step`, which would otherwise treat a
    /// non-Measuring committed's raw as evidence on every 1 s tick, violating decision 11).
    /// A no-op unless `committed` is currently `.measuring`.
    @discardableResult
    public func commitFromMeasuring(_ raw: QuotaState, monoMs: Int64) -> QuotaState {
        guard committed == .measuring else { return committed }
        commitImmediately(raw, monoMs: monoMs)
        return committed
    }

    private func resetPendingEvidence() {
        escalateHistory.removeAll()
        pendingDeescalateState = nil
        pendingDeescalateCount = 0
        pendingDeescalateSinceMono = nil
    }

    private func commitImmediately(_ state: QuotaState, monoMs: Int64) {
        committed = state
        lastChangeMono = monoMs
        resetPendingEvidence()
    }
}
